chore : added poetry for package management

vj-codes opened the pull request:
### Description
This PR is a continuation of #914 which uses Poetry has the default dependency management tool

Fixes #691 

### Type of Change:

- Code
- Quality Assurance
- Documentation

### How Has This Been Tested?

### Checklist:

<!-- **Delete irrelevant options.** -->

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**

- [ ] My changes generate no new warnings 
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

bartekpacia committed:
feat: use Poetry for package management

bartekpacia committed:
Merge branch 'develop' into use_poetry_new

bartekpacia committed:
improve README

bartekpacia committed:
update requirements.txt

bartekpacia committed:
Merge branch 'develop' into use_poetry_new

bartekpacia committed:
Update README.md

Co-authored-by: Isabel Costa <11148726+isabelcosta@users.noreply.github.com>

vj-codes committed:
resolved conflicts from poetry PR

vj-codes committed:
resolved conflicts

vj-codes commented:
@isabelcosta hey I have resolved the conflicts from the previous PR , as expected all commits have been included in this PR, and about `poetry add psycopg2` I need to install poetry first 👀 but all that's need to be done is to use the command and commit right?

epicadk commented on README.md:
````diff
@@ -191,7 +190,7 @@ docker run --env "FLASK_APP=run.py" --publish 5000:5000 mentorship-backend:lates
 
 ```
-python -m unittest discover tests
+poetry run test
````
Make changes to github actions

epicadk commented on Dockerfile:
```diff
@@ -1,6 +1,6 @@
 RUN pip install --no-cache-dir -r requirements.txt
 COPY . /dockerBuild
-CMD ["flask", "run", "--host", "0.0.0.0"]
+CMD ["poetry", "run", "start"]
```
Let this be. Docker is containerized anyways.

isabelcosta replied:
I guess, that `poetry start` runs only `python run.py`, and that isn't what was originally in the Dockerfile, so it will probably change the behavior of how this works 🤔

epicadk reviewed (changes requested):
Thankyou, please have a look at the comments

isabelcosta commented on README.md:
```diff
@@ -5,6 +5,9 @@
 [![project chat](https://img.shields.io/badge/zulip-join_chat-brightgreen.svg)](https://anitab-org.zulipchat.com/#narrow/stream/222534-mentorship-system)
 [![Python 3.x](https://img.shields.io/badge/python-3.x-blue.svg)](https://www.python.org/downloads/)
 [![Docs](https://img.shields.io/badge/documentation-mentorship--backend-blue.svg)](https://anitab-org.github.io/mentorship-backend)
+=======
+[![Python 3.9](https://img.shields.io/badge/python-3.9-blue.svg)](https://www.python.org/downloads/)
+>>>>>>> 8f044bccb816f8a07789312d8991d8d8d5cbed0b
```
```suggestion
[![Python 3.9](https://img.shields.io/badge/python-3.9-blue.svg)](https://www.python.org/downloads/)
```
rests from merge conflict markup :P

vj-codes committed:
made requested changes

vj-codes committed:
made changes to gh actions

vj-codes commented:
@isabelcosta @epicadk I've added psycopg2 and made the requested changes , 
`poetry: command not found` this error is shown when I updated gh actions to run tests with poetry so I reverted the change
After using poetry locally here are my inputs-
1) poetry run start and poetry run tests works as expected
2) poetry shell probably creates a virtual env but the word `venv` does not appear like it used to before so how to confirm the activation?
3) And what are those codecov warnings about on github?

![Screenshot (3)](https://user-images.githubusercontent.com/60894542/116823162-9356fa00-aba0-11eb-9565-35ae3fe25f62.png)

isabelcosta commented on README.md:
````diff
@@ -51,14 +50,11 @@ To setup the project locally read these wiki pages and follow the instructions:
-```
+1. [Install Poetry](https://python-poetry.org/docs/#installation):
 
 2. Activate the virtual environment:
````
With step 1 being about poetry, does this step 2 make sense? Since as you said `venv` folder is not a thing?

isabelcosta commented on README.md:
````diff
@@ -98,9 +92,11 @@ export DB_ENDPOINT=<database_endpoint>
+poetry run start
 ```
 
 6. Navigate to http://localhost:5000 in your browser
````
a few lines below (i can't put a comment there), there are instructions for running on Linux, was that part supposed to go away? Since you removed the entry on the readme table of contents?

vj-codes replied:
@isabelcosta since the virtual env is being done by poetry the command would be same on both linux and windows I think. It needs to be confirmed though 👀

epicadk commented on pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+[tool.poetry]
+name = "mentorship-backend"
```
Why is this required?

vj-codes commented on pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+readme = "README.md"
+homepage = "https://github.com/anitab-org/mentorship-backend"
+repository = "https://github.com/anitab-org/mentorship-backend"
+documentation = "https://github.com/anitab-org/mentorship-backend/wiki"
```
Should this link be updated to the docusaurus one?

vj-codes commented:
The label was still on needs review due to which I forgot there were changes requested, will update them and resolve the conflict in a day!

vj-codes committed:
made reuqested changes

vj-codes committed:
Merge branch 'develop' into use_poetry_new

vj-codes committed:
resolved conflicts 2

vj-codes committed:
resolved conflicts 3

isabelcosta commented:
Ok so to **test** this, we should:
1. Clone the repo
2. Set it up from scratch like a newcomer
3. Check if the setup was successful

**Goal:** We need to check that all the necessary instructions on README are enough to set this successfully as a newcomer would.

cc @anitab-org/mentorship-maintainers

devkapilbansal commented:
> Ok so to **test** this, we should:
> 
> 1. Clone the repo
> 2. Set it up from scratch like a newcomer
> 3. Check if the setup was successful
> 
> **Goal:** We need to check that all the necessary instructions on README are enough to set this successfully as a newcomer would.
> 
> cc @anitab-org/mentorship-maintainers

Hi,
I setup the repository locally and this are the problems I found.
1. I have to setup the PATH manually. It is written in install instructions that path will be set automatically but that didn't happen in my case.
2. I followed the steps mentioned in README but have to navigate to last for installing the dependencies. I would suggest adding the section link after Installing poetry.
3. I got port already in use error when I tried running the app. I changed it to 4000 then and app worked normally. Although I am not sure if this was a problem in the code or with my system only.

vj-codes commented:
@devkapilbansal what PATH are you speaking about and which OS?

devkapilbansal commented:
> @devkapilbansal what PATH are you speaking about and which OS?

@vj-codes path to Poetry in environment. I tried it on Ubuntu 20.04

vj-codes commented:
@devkapilbansal in my case it did get set automatically on windows though
I agree with moving the commands to keep them in the same place 👍

vj-codes committed:
moved poetry commands and removed run app in linux part

Removed run app in Linux as it was same as instructions to run locally

vj-codes commented:
@devkapilbansal I've moved the commands and about the path we can't do anything as it's a part of poetry docs and not here 
@gaurivn can you test it please and share if you also have the same issue? 
Or maybe we can add something like if the commands don't work after installing try setting up the path manually, what say?

devkapilbansal commented:
@vj-codes runnimg this command do the needful for me
`export PATH=$HOME/.poetry/bin:$PATH`

We can add this if in case poetry doesn't work after installation. What's say?

vj-codes committed:
Update README.md

vj-codes committed:
Merge branch 'develop' into use_poetry_new

vj-codes committed:
Update requirements.txt

vj-codes committed:
Merge branch 'develop' into use_poetry_new

isabelcosta reviewed (approved):
Great work @vj-codes 🎉
Will test this and let you all know here :)

devkapilbansal reviewed (commented):
Hi @vj-codes 
I wonder why we need to maintain a `requirements.txt` file with poetry. The dependencies are already defined in `pyproject.toml` file and if we still need to manage requirements file then what's the benefit of using poetry?
If I am not wrong, we decided to use poetry for package management because of a lot of conflicts in dependencies :thinking:

isabelcosta committed:
Merge branch 'develop' into use_poetry_new
